libyul/AsmParser.cpp

```diff
@@ -512,14 +512,24 @@ variant<Literal, Identifier> Parser::parseLiteralOrIdentifier()
 	case Token::FalseLiteral:
 	{
 		LiteralKind kind = LiteralKind::Number;
+		string currLiteral = currentLiteral();
+
 		switch (currentToken())
 		{
 		case Token::StringLiteral:
 		case Token::HexStringLiteral:
 			kind = LiteralKind::String;
 			break;
 		case Token::Number:
-			if (!isValidNumberLiteral(currentLiteral()))
+			currLiteral.erase(std::remove(currLiteral.begin(), currLiteral.end(), '_'), currLiteral.end());
+			if (currLiteral.find("e") != string::npos && !(boost::starts_with(currLiteral, "0x") || boost::starts_with(currLiteral, "0X")))
+			{
+				istringstream os(currLiteral);
+				double d;
+				os >> d;
+				currLiteral = to_string(d).substr(0, to_string(d).find('.'));
+			}
```

Comment:
This is not the right way to go about it. Solidity and Yul share the scanner class and it already supports underscores, scientific notation, hexadecimals - everything that is allowed in number literals at Solidity level. The function for getting the next token calls `Scanner::scanNumber()` when it encounters a number literal, and that's where the magic happens: https://github.com/ethereum/solidity/blob/4c1cadb7160a48869ba50ae0cb9beae2c6bc553b/liblangutil/Scanner.cpp#L920-L1012

So you don't need to implement this functionality from scratch. It's already there. What you actually need to do is to relax the validation in `isValidNumberLiteral()` so that it does not reject numbers with underscores as invalid.

Reply:
I'm getting the following Boost error when I relax the validation in isValidNumberLiteral. It seems to be coming from when the string with underscores in get's passed to `YulString{}` in `parseLiteralOrIdentified`

```
Uncaught exception:
/usr/include/boost/multiprecision/cpp_int.hpp(1781): Throw in function void boost::multiprecision::backends::cpp_int_backend<MinBits, MaxBits, SignType, Checked, Allocator>::do_assign_string(const char*, const false_&) [with unsigned int MinBits = 0; unsigned int MaxBits = 0; boost::multiprecision::cpp_integer_type SignType = boost::multiprecision::signed_magnitude; boost::multiprecision::cpp_int_check_type Checked = boost::multiprecision::unchecked; Allocator = std::allocator<long long unsigned int>; mpl_::false_ = mpl_::bool_<false>]
Dynamic exception type: boost::wrapexcept<std::runtime_error>
std::exception::what: Unexpected character encountered in input.
```

Reply:
Yeah, well, the scanner gives you the literal as a string. It still has to be parsed if you want to get the value. Yul was designed to be super simple so until now you could not do anything fancy with literals. This way we were able to get away with simply passing them to `u256` constructor and it would parse the number for us. In Solidity that's not the case and we instead have to use `RationalNumberType::literalValue()`, which does the parsing whenever we need the value.

You can't use that in Yul obviously (`RationalNumberType` is a high-level Solidity type and only available in libsolidity) so you have to add a function for parsing the number. Which here simply means stripping the underscores and passing the result to `u256`.

Reply:
I've removed the underscores from the number literal in `parseLiteralOrIdentified` so that just the number is passed to YulString{}. Do you want the underscores to be removed in another location?

The only reason I haven't removed the underscores in `IsValidNumberLiteral` is because `YulString{}` and `IsValidNumberLiteral` are both called from `parseLiteralOrIdentified` so it made sense to me to make the change in `parseLiteralOrIdentified` before either method was called.

Reply:
`parseLiteralOrIdentifier()` is not the right place to do the stripping because if you do that, the stripped version will be stored in the AST. This may break some things. E.g. the string will no longer match the size of the source location stored in the AST as well. You should store the raw, unparsed value in the AST just as it was until now and parsing it into value should happen later.

Reply:
Where do you think we should do the parsing?

Reply:
Just like in Solidity, at the point where the value is accessed.

Before that though, we really need to decide if we want this feature in Yul at all. The issue didn't really go through any design discussion and was just quickly triaged, without a proper decision to proceed with it.

It sounds like a good idea but Yul is meant more for tools and the simplicity of the grammar is also a very important factor.